server.js

```diff
@@ -14,10 +12,103 @@ const welcomeMessage = {
+app.use(bodyParser.urlencoded({ extended: false }));
+
 const messages = [welcomeMessage];
 
 app.get("/", function (request, response) {
   response.sendFile(__dirname + "/index.html");
+  response.status(200).send("Welcome to my message App");
 });
 
-app.listen(process.env.PORT);
+/*to get all the messages*/
+app.get("/messages", function (request, response) {
+  response.status(200).send({ messages });
+});
+
+/*to post a message*/
+app.post("/messages", function (req, res) {
+  const newMessage = req.body;
+  /*access Id */
+  const previousMessage = messages[messages.length - 1];
+  const currentId = previousMessage.id + 1;
+
+  /*merged objects in the final one*/
+  const newMessageWithId = { ...newMessage, id: currentId };
+
+  /*validate message text and from properties*/
+  if (!newMessageWithId.text || !newMessageWithId.from) {
+    return res.status(400).json({
+      error: "Both name and message are required",
+    });
+  }
+
+  messages.push(newMessageWithId);
+  res.status(201).send({ newMessageWithId });
+});
+
+/*To get by ID*/
+app.get("/messages/:id", function (request, response) {
+  // console.log(request.params, "<------- req params");
+  const idToFind = Number(request.params.id);
+  const message = messages.find((message) => message.id === idToFind);
+  console.log(message);
+  response.status(200).send({ message });
+});
+
+/*to delete a message by ID */
+app.delete("/messages/:id", function (request, response) {
+  const idToFind = Number(request.params.id);
+  const message = messages.find((message) => message.id === idToFind);
+  const indexOfMessage = messages.indexOf(message);
+
+  if (indexOfMessage !== -1) {
+    messages.splice(indexOfMessage, 1);
+    response
+      .status(200)
+      .send({ message: `Message with ID ${indexOfMessage} has been deleted` });
+  } else {
+    response
+      .status(404)
+      .send({ message: `Message with ID ${indexOfMessage} not found` });
+  }
+
+  // response.status(200).send({ message });
+});
+
+/*to search messages by text*/
+app.get("/messages/search", function (req, res) {
+  const searchText = req.query.text;
+  const matchingMessages = messages.filter(
+    (message) => message.text.indexOf(searchText) !== -1
+  );
+  res.status(200).send({ messages: matchingMessages });
+});
+
+/*to get the latest 10 messages*/
+app.get("/messages/latest", function (req, res) {
+  const latestMessages = messages.slice(-10);
+  res.status(200).send({ messages: latestMessages });
+});
+
+app.put("/messages/:id", function (req, res) {
+  const id = parseInt(req.params.id); // get the ID from the URL parameters and convert it to a number
+
+  const newMessage = { ...req.body, id: Number(req.params.id) };
+  const messageIndex = messages.findIndex((q) => q.id === Number(req.params.id));
+  messages.splice(messageIndex, 1, newMessage);
+  res.status(200).send({ success: true });
+});
+
+// const PORT = 9090;
+// app.listen(process.env.PORT);
+
+app.listen(9090, () => {
```

Comment:
you can listen to port specified by `process.env.port` when it is present or port 9090 by saying
`app.listen( process.env.port || 9090, `
Again, its better to use a variable to store the default port number 9090

Reply:
got it, thanks for reviewing the code, Jack!!

Reply:
I dont have definite answers to your question but just want to highlight the importance of 
1. Input validation
2. Error handling
I think Alun's material provided examples on the above aspects. You may take a look